flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunction.java

```diff
@@ -0,0 +1,381 @@
+		try {
+			bufferedResultsLock.lock();
+			if (bufferedResults.size() >= maxResultsBuffered) {
+				bufferNotFullCondition.await();
+			}
+			bufferedResults.add(value);
+		} finally {
+			bufferedResultsLock.unlock();
+		}
+	}
+
+	@Override
+	public void close() throws Exception {
+		serverThread.close();
+	}
+
+	public void accumulateFinalResults() throws Exception {
+		try {
+			bufferedResultsLock.lock();
+
+			// put results not consumed by the client into the accumulator
+			// so that we do not block the closing procedure while not throwing results away
+			SerializedListAccumulator<IN> listAccumulator = new SerializedListAccumulator<>();
+			for (IN result : bufferedResults) {
+				listAccumulator.add(result, serializer);
+			}
+			LongCounter offsetAccumulator = new LongCounter(firstBufferedResultOffset);
+			getRuntimeContext().addAccumulator(finalResultListAccumulatorName, listAccumulator);
+			getRuntimeContext().addAccumulator(finalResultOffsetAccumulatorName, offsetAccumulator);
+		} finally {
+			bufferedResultsLock.unlock();
+		}
+	}
+
+	public void notifyCheckpointComplete(long checkpointId) {
+		this.lastCheckpointId = checkpointId;
+	}
+
+	public void setOperatorEventGateway(OperatorEventGateway eventGateway) {
+		this.eventGateway = eventGateway;
+	}
+
+	/**
+	 * The thread that runs the socket server.
+	 */
+	private class ServerThread extends Thread {
+
+		private final ServerSocket serverSocket;
+
+		private boolean running;
+
+		private Socket connection;
+		private DataInputViewStreamWrapper inStream;
+		private DataOutputViewStreamWrapper outStream;
+
+		private ServerThread() throws Exception {
+			this.serverSocket = new ServerSocket(0, 0, getBindAddress());
+			this.running = true;
+		}
+
+		@Override
+		public void run() {
+			while (running) {
+				try {
+					if (connection == null) {
+						// waiting for coordinator to connect
+						connection = serverSocket.accept();
+						inStream = new DataInputViewStreamWrapper(this.connection.getInputStream());
+						outStream = new DataOutputViewStreamWrapper(this.connection.getOutputStream());
+						LOG.info("Coordinator connection received");
+					}
+
+					CollectCoordinationRequest request = new CollectCoordinationRequest(inStream);
+					String requestVersion = request.getVersion();
+					long requestOffset = request.getOffset();
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(
+							"Request received, version = " + requestVersion + ", offset = " + requestOffset);
+						LOG.debug(
+							"Expecting version = " + version + ", firstBufferedOffset = " + firstBufferedResultOffset);
+					}
+
+					if (!version.equals(requestVersion) || requestOffset < firstBufferedResultOffset) {
+						// invalid request
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Invalid request");
+						}
+						sendBufferedResults(0, 0);
+						continue;
+					}
+
+					// valid request, sending out results
+					try {
+						bufferedResultsLock.lock();
+
```

Comment:
add some comment about the `requestOffset` here. Like it means client already has all results with offset less than  `requestOffset`, so we can safely drop them from buffer